src/common/container_helpers.h

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2022, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Miscellaneous container helpers.
+
+#pragma once
+
+//local headers
+
+//third party headers
+
+//standard headers
+#include <algorithm>
+#include <functional>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+//forward declarations
+
+
+namespace tools
+{
+
+/// convert a binary comparison function to a functor
+template <typename T, typename ComparisonOpT = bool(const T &a, const T &b)>
+inline auto compare_func(const ComparisonOpT &comparison_function)
+{
+    return [&comparison_function](const T &a, const T &b) -> bool { return comparison_function(a, b); };
```

Comment:
I need this adaptor to disambiguate function overloads at the callsite. [Example of adaptor](https://github.com/UkoeHB/monero/blob/3517c428dd1232cdf14dda9b4d74fe0eb6b1bbe2/src/seraphis/tx_builders_mixed.cpp#L857), [example of is_sorted_and_unique()](https://github.com/UkoeHB/monero/blob/3517c428dd1232cdf14dda9b4d74fe0eb6b1bbe2/src/seraphis/tx_validators.cpp#L356).

Reply:
However, it's probably fine to update the adaptor to return a `std::cref()`.

Reply:
> disambiguate function overloads at the callsite.

Yikes. There has to be a built-in way to handle this. It's easy with templated overloads, but not certain (off-top-of-head) how with non-templated functions.

Anyway, yes, take the argument by-value and then do the lambda capture by-value. Prevents any unexpected nastiness if someone uses the function in some funky way too.

Reply:
And it could be a C++14 move-capture-by-value too.

Reply:
Just remembered the typical C++ way to handle this -
```c++
struct compare_KI_
{
    bool operator()(const key_image1&, const key_image2&) const;
    bool operator()(const key_image2&, const key_image2&) const;
};
constexpr const compare_KI_ compare_KI{};
```
And then the overload selection is chosen inside `std::sort` instead of at the call-site to `std::sort`.  But this technique you've come up with has a bit less boilerplate and doesn't require the declarations to be in the same header.